include/envoy/matcher/matcher.h

```diff
@@ -0,0 +1,154 @@
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "envoy/common/pure.h"
+#include "envoy/config/common/matcher/v3/matcher.pb.h"
+#include "envoy/config/typed_config.h"
+
+#include "absl/strings/string_view.h"
+#include "absl/types/optional.h"
+#include "external/envoy_api/envoy/config/core/v3/extension.pb.h"
+
+namespace Envoy {
+namespace Matcher {
+
+// This file describes a MatchTree<DataType>, which traverses a tree of matches until it
+// either matches (resulting in either an action or a new tree to traverse) or doesn't match.
+// The matching might stop early if either the data is not available at all yet, or if more data
+// might result in a match.
+
+// By returning a new tree when an OnMatch results in a new tree, matching can be resumed from
+// this tree should more data be required to complete matching. This avoids having to start
+// from the beginning every time. At some point we might support resuming for any node in the match
+// tree: this requires some careful handling of tracking which on_no_match to use should we fail to
+// match.
+//
+// All the matching is performed on strings: a DataInput<DataType> is used to extract a specific
+// string from an instance of DataType, while an InputMatcher is used to determine whether the
+// extracted string is a match.
+//
+// For example, DataType might be the type HttpDataInput, allowing
+// for the use of HttpRequestHeaders : DataInput<HttpDataInput>, which is configured with the name
+// of the header to extract from the request headers.
+//
+// In cases where the data to match on becomes available over time, this would be fed into the
+// DataType over time, allowing matching to be re-attempted as more data is made available. As such
+// whenever we extract data from a DataInput, we make note of whether the data might change and
+// pause matching until we either match or have all the data. It would then fall on the caller to
+// both provide more information to the DataType and to resume matching.
+template <class DataType> class MatchTree;
+
+template <class DataType> using MatchTreeSharedPtr = std::shared_ptr<MatchTree<DataType>>;
+
+using TypedExtensionConfigOpt = absl::optional<envoy::config::core::v3::TypedExtensionConfig>;
+
+// On match, we either return the action to perform or another match tree to match against.
+template <class DataType> struct OnMatch {
+  const TypedExtensionConfigOpt action_;
```

Comment:
Don't you want this to be pre-resolved to a concrete impl or factory for an impl? Otherwise aren't we going to pay a conversion/parse tax on every match?

Reply:
That's a good point, I'll add in another interface for this